### PR TITLE
Various zfs block mode fixes

### PR DESCRIPTION
**Pull request by monstermunchkin**
Fixes https://github.com/lxc/lxd/issues/11308

### CODE REVIEW

**Comment by tomponline** [comment]
Tests failing

**Comment by monstermunchkin** [comment]
jenkins: test this please

File: test/suites/migration.sh
Comment (tomponline) [inline test/suites/migration.sh:186]: I'd like to know more about this. As it feels like we're testing for bug here. Does this same issue exist for LVM (with xfs)?

Comment (tomponline) [reply test/suites/migration.sh:186]: Which was the operation that would have triggered the unmount we are waiting for here? Shouldn't that not return until the unmount has finished?

File: lxd/storage/drivers/driver_btrfs.go
Comment (tomponline) [inline lxd/storage/drivers/driver_btrfs.go:452]: Shouldnt this be in this PR https://github.com/lxc/lxd/pull/11368?

File: lxd/storage/drivers/driver_zfs.go
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs.go:596]: Should this be in https://github.com/lxc/lxd/pull/11368?

File: lxd/storage/drivers/driver_btrfs.go
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_btrfs.go:452]: Originally, I wanted to have 2 separate PRs (one for rsync, the other for various zfs block mode fixes) but I noticed they needed to be merged in order for all tests to succeed. I'll be closing the other one in favour of this one.

File: test/suites/migration.sh
Comment (monstermunchkin) [reply test/suites/migration.sh:186]: > Does this same issue exist for LVM (with xfs)?

I don't know.

> Which was the operation that would have triggered the unmount we are waiting for here? Shouldn't that not return until the unmount has finished?

I investigated this, and found that the following code snippet fails:
```
  echo test | lxc_remote file push - l1:c1/tmp/foo
  lxc_remote copy l1:c1 l2:c1
```

This seems to be caused by forkfile still running while attempting to copy. @stgraber proposed [this fix](https://gist.github.com/stgraber/4133fe08df647b9fac4ce3d96b8a7b22) which I'm currently testing.

**Comment by monstermunchkin** [comment]
Tests should succeed once https://github.com/lxc/lxd/pull/11419 is merged.

File: lxd/storage/drivers/driver_zfs_volumes.go
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2677]: It feels like this change for unix.MS_RDONLY and presumably the ext4 bits below should be a separate commit from the splitting of the function (and the reverter being added), accompanied by a commit message explaining *why* its needed.

File: lxd/storage/drivers/driver_zfs_volumes.go
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2810]: I'd like to see this as a separate commit, along with explanation as to what its fixing. As I've taken a look at the LVM Ceph drivers and they don't have this behaviour.

File: lxd/storage/drivers/driver_zfs_volumes.go
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:3026]: Shouldnt there be a deactivate at some point here?

File: lxd/storage/drivers/driver_zfs_volumes.go
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2763]: Why is this needed?

File: lxd/storage/drivers/driver_btrfs.go
Comment (tomponline) [reply lxd/storage/drivers/driver_btrfs.go:462]: It would be good to have the rsync one split from this PR. It makes it easier to review and makes it possible to make progress more quickly. I'm not really clear why you can't base this PR off of the separate one for the rsync changes, and then once its merged you can rebase.

File: doc/api-extensions.md
Comment (tomponline) [inline doc/api-extensions.md:2154]: When used in backticks, the rest of the documentation uses `btrfs` and `xfs` (lower case).

@ru-fu what are your thoughts on this?

Perhaps rephrasing to say just:

> This adds support for using ZFS block filesystem volumes allowing the use different filesystems on top of ZFS.

Comment (ru-fu) [reply doc/api-extensions.md:2154]: We've been using Btrfs and XFS in the docs, since those are clear abbreviations. "ext4" is a bit of a special case there, since it contains a number and must be lowercase. So I would go for "Btrfs, `ext4`, and XFS".

I'm fine with your suggestion as well except that "file system" in general is two words - apart from "ZFS filesystem", which is their terminology. \*argh*
So that would be:
```
This adds support for using ZFS block {spellexception}`filesystem` volumes allowing the use of different file systems on top of ZFS.
```

Comment (tomponline) [reply doc/api-extensions.md:2154]: Thanks!

File: lxd/storage/drivers/driver_zfs_volumes.go
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go:2763]: `tmpVol` is never actually created so running `d.DeleteVolume(tmpVol, op)` won't do anything. What we do is run `zfs clone <snapshotDataset> <dataset>` which we need to clean up.

Comment (tomponline) [reply lxd/storage/drivers/driver_zfs_volumes.go:2763]: OK I think I'll hold off reviewing this further until this PR is split out a bit, as it feels like this PR is trying to address too many things at this time and its tricky to see how its all fitting together. Thanks.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by monstermunchkin** [comment]
@stgraber this needs rebasing, and some commits need to be removed. Please don't merge yet.

**Comment by monstermunchkin** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
@monstermunchkin want a review on this yet?

**Comment by monstermunchkin** [comment]
@tomponline yes, now you can. I just had to fix a migration issue.

File: lxd/storage/drivers/driver_zfs_volumes.go
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2161]: We only need to do this if volSrcArgs.AllowInconsistent is false. So need a check for it here.

Similar to https://github.com/lxc/lxd/blob/e19b4dfd1f571d5466000b9c8601017dd155280a/lxd/storage/drivers/driver_zfs_volumes.go#L578

**Review by tomponline — CHANGES_REQUESTED** [review]
Just one change needed :)

File: lxd/storage/drivers/driver_zfs_volumes.go
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go]: Right. It's fixed.

**Pull request merged**
